packages/medusa/src/utils/transaction/transaction-builder.ts

```diff
@@ -0,0 +1,358 @@
+import { TransactionStepsDefinition } from "./types"
+
+export type ActionHandler = {
+  [type: string]: (data: any, context: any) => Promise<any>
+}
+
+interface InternalStep extends TransactionStepsDefinition {
+  next?: InternalStep | InternalStep[]
+  depth: number
+  parent?: InternalStep | null
+}
+
+export class TransactionBuilder {
+  private steps: InternalStep
+
+  constructor(steps?: TransactionStepsDefinition) {
+    this.load(steps)
+  }
+
+  load(steps?: TransactionStepsDefinition) {
+    this.steps = {
+      depth: -1,
+      parent: null,
+      next: steps
+        ? JSON.parse(
```

Comment:
question: Why do we stringify and parse here?

Reply:
Deep copying the object to do not mess with the original one.

Reply:
Not sure it works well with some function or object type no?

Reply:
we only have primitive types describing the transaction steps.

Reply:
but we have the functions included in the steps right?